session.go

```diff
@@ -1729,8 +1697,73 @@ func (b *Batch) attempt(keyspace string, end, start time.Time, iter *Iter, host
 }
 
 func (b *Batch) GetRoutingKey() ([]byte, error) {
-	// TODO: use the first statement in the batch as the routing key?
-	return nil, nil
+	if b.routingKey != nil {
+		return b.routingKey, nil
+	}
+
+	if len(b.Entries) == 0 {
+		return nil, nil
+	}
+
+	entry := b.Entries[0]
+	if entry.binding != nil {
+		// bindings do not have the values let's skip it like Query does.
+		return nil, nil
+	}
+	// try to determine the routing key
+	routingKeyInfo, err := b.session.routingKeyInfo(b.Context(), entry.Stmt)
+	if err != nil {
+		return nil, err
+	}
+
+	return b.routingKeyCreator.createRoutingKey(routingKeyInfo, entry.Args)
+}
+
+type routingKeyCreator struct {
```

Comment:
I think we can just get rid of this type, saving this byte allocation wont make a difference if people dont reuse the batch/query.

Reply:
There, the type removed and the method is a function. Should there perhaps be a pool of `[]byte` for the buffer?